tests/h/views/badge_test.py

```diff
@@ -1,91 +1,96 @@
 # -*- coding: utf-8 -*-
 
+from datetime import datetime
 from unittest import mock
 
 import pytest
-from h_matchers import Any
 from pyramid import httpexceptions
 from webob.multidict import MultiDict
 
-from h.views.badge import badge
+from h.views.badge import Blocklist, badge
 
-badge_fixtures = pytest.mark.usefixtures("models", "search_lib")
 
+class TestBlocklist:
+    @pytest.mark.parametrize("bad_part", Blocklist.BLOCKED_URL_PARTS)
+    @pytest.mark.parametrize("prefix", ("http:", "https:"))
+    def test_it_blocks(self, bad_part, prefix):
+        url = f"{prefix}{bad_part}/path?a=b"
 
-@badge_fixtures
-def test_badge_returns_number_from_search(
-    models, pyramid_request, search_run, mark_uri_as_annotated
-):
-    mark_uri_as_annotated("http://example.com")
+        assert Blocklist.is_blocked(url)
 
-    pyramid_request.params["uri"] = "http://example.com"
-    models.Blocklist.is_blocked.return_value = False
-    search_run.return_value = mock.Mock(total=29)
+    def test_it_allows_non_blocked_items(self):
+        assert not Blocklist.is_blocked("http://example.com/this/is/fine")
 
-    result = badge(pyramid_request)
+    def test_its_fast(self):
+        # Check any modifications haven't made this significantly slower
+        reps = 10000
 
-    search_run.assert_called_once_with(
-        MultiDict({"uri": "http://example.com", "limit": 0})
-    )
-    assert result == {"total": 29}
+        start = datetime.utcnow()
+        for _ in range(reps):
+            Blocklist.is_blocked("http://example.com/this/is/fine")
 
+        diff = datetime.utcnow() - start
 
-@badge_fixtures
-def test_badge_does_not_search_if_uri_never_annotated(
-    models, pyramid_request, search_run
-):
-    pyramid_request.params["uri"] = "http://example.com"
-    models.Blocklist.is_blocked.return_value = False
+        seconds = diff.seconds + diff.microseconds / 1000000
+        calls_per_second = int(reps // seconds)
+        print(
+            f"Calls per second: {calls_per_second}, {1000000 / calls_per_second:.03f} μs/call"
```

Comment:
I suspect the most likely cause of a slowdown in the badge would be a modification to any of the machinery or middleware that is common to all requests, rather than the `Blocklist` class. For that reason I'd probably look to rely mainly on at system-level detection of regressions in badge transaction performance in NR.

Reply:
This won't help you detect reasons for slowness outside of this class, but as a unit test it's not intended to. I just wanted to make sure our "quick" optimisation step isn't accidentally made very slow.